Improve PICSAR QED library structure and CMakeLists

lucafedeli88 opened the pull request:
This PR started as an improvement of CMakeLists to make the integration of the QED library in WarpX easier. However, it ended up as a large refactoring  of the whole QED library. Below I list all the main changes. I would really like to thank @ax3l for his help during the preparation of this PR!

- The library files have been moved from `picsar/src/multi_physics/QED/src/` to `picsar/multi_physics/QED/include/picsar_qed`
- All other files (e.g. tests) are now in `picsar/multi_physics/QED`
- There are automated unit tests on Linux and macOS (thanks @ax3l )!
- CMakeLists files have been greatly improved (thanks @ax3l )
- Relative include paths in library files have been replaced with full paths starting from `picsar_qed` to avoid collisions (thanks @ax3l )
- By default Boost special functions are used (instead of those provided by the C++17 standard library). This is because I've found out that Boost implementations behave better for large arguments.
- A fix was introduced to be able to compile with `DCP++` (SYCL), since it does not support `std::floor`
- README file has been updated and some minor bugs have been corrected
- `INTERNAL` preprocessor variables were removed
- Added `codeQL` analysis (thanks @ax3l )


**warning:** do not merge! Merging must be done simultaneously with https://github.com/ECP-WarpX/WarpX/pull/1529

lucafedeli88 committed:
improved CMakeLists

lucafedeli88 committed:
updated another CMakeLists

lucafedeli88 committed:
fixed minor issues

lucafedeli88 committed:
fixed issue

lucafedeli88 committed:
fixed bug

ax3l commented on src/multi_physics/CMakeLists.txt:
```diff
@@ -1,13 +1,33 @@
+  HOMEPAGE_URL "https://github.com/ECP-WarpX/picsar"
+  LANGUAGES CXX)
+  
+add_library(${PROJECT_NAME} INTERFACE)
```
Adding this alias will make `add_subdirectory` and `find_package` work with the same name:
```suggestion
add_library(${PROJECT_NAME} INTERFACE)
add_library(${PROJECT_NAME}::${PROJECT_NAME} ALIAS ${PROJECT_NAME})
```

ax3l commented on src/multi_physics/QED/CMakeLists.txt:
```diff
@@ -1,19 +1,14 @@
+find_package(Boost)
+if(Boost_FOUND)
+    include_directories(${Boost_INCLUDE_DIRS})
+    # TODO: define a preprocessor variable to let the library know if Boost is available or not
```
`target_compile_definitions(pxrmp_qed PUBLIC PXRMP_HAS_BOOST=1)`

ax3l commented on src/multi_physics/QED_tests/CMakeLists.txt:
```diff
@@ -1,8 +1,7 @@
 #Check if unit_test package is available
 find_package(Boost COMPONENTS unit_test_framework REQUIRED)
-include_directories (${Boost_INCLUDE_DIRS})
+include_directories(${Boost_INCLUDE_DIRS})
+add_definitions(-DBOOST_TEST_DYN_LINK)
```
```suggestion
```

ax3l replied:
Includes, defined and libs added on-target with `target_link_libraries`:
```suggestion
```

ax3l commented on src/multi_physics/QED_tests/CMakeLists.txt:
```diff
@@ -1,8 +1,7 @@
-cmake_minimum_required(VERSION 3.5)
-
 #Check if unit_test package is available
 find_package(Boost COMPONENTS unit_test_framework REQUIRED)
```
```suggestion
find_package(Boost <MIN_VERSION> COMPONENTS unit_test_framework REQUIRED)
```

ax3l commented on src/multi_physics/QED_tests/CMakeLists.txt:
```diff
@@ -23,11 +22,7 @@ foreach(test_src ${TEST_SRCS})
-        else()
-          set_target_properties(${test_name} PROPERTIES CXX_STANDARD 14)
-        endif()
+        target_compile_features(${test_name} PUBLIC cxx_std_14)
```
```suggestion
        target_compile_features(${test_name} PUBLIC cxx_std_14)
        target_link_libraries(${test_name} PRIVATE Boost::unit_test_framework)
```

ax3l commented on src/multi_physics/QED_table_generator/CMakeLists.txt:
```diff
@@ -9,7 +9,7 @@ set(
     )
 
-include_directories(../QED)
+include_directories(../.)
```
The `.` looks confusing and should be superfluous.

lucafedeli88 replied:
Here I wanted to be able to compile QED table generator separately (e.g. being able to do `mkdir build ; cd build ; cmake ../.` directly from `QED_table_generator` directory. If doing this is bad for some reason we can change it.

ax3l replied:
I think we forgot to come back to this yesterday.

If you want to build it from `src/multi_physics/QED_table_generator/` then you can hack it like this:
```cmake
target_include_directories(qed_table_generator PRIVATE ..)
target_include_directories(qed_table_generator PRIVATE .)
```

Does it really depend on all of `multi_physics/` or only on things in `QED/`? In that case I would move `QED_table_generator/` to `QED/table_generator/` and do: `target_link_libraries(qed_table_generator PRIVATE PXRMP::QED)`.
Then `add_subdirectory` the `table_generator` in `QED/CMakeLists.txt` and build it with QED - potentially set controlled via an option such as `option(PXRMP_QED_TOOLS ...`).

lucafedeli88 replied:
Thanks for the advice! I should have followed it: I moved `QED_table_generator` into `QED` and I have added the option `PXRMP_QED_TOOLS` .

lucafedeli88 replied:
However, Ubuntu tests are failing with this weird message:
```
Err:20 https://packagecloud.io/github/git-lfs/ubuntu focal InRelease
  402  Payment Required [IP: 54.183.38.243 443]
```

ax3l commented on src/multi_physics/QED/CMakeLists.txt:
```diff
@@ -1,19 +1,14 @@
-target_include_directories(pxrmp_qed INTERFACE src/physics/schwinger)
-target_include_directories(pxrmp_qed INTERFACE src/physics/breit_wheeler)
-target_include_directories(pxrmp_qed INTERFACE src/physics/quantum_sync)
+find_package(Boost)
```
```suggestion
find_package(Boost 1.65.1 COMPONENTS math)
```

ax3l commented on src/multi_physics/QED/CMakeLists.txt:
```diff
@@ -1,19 +1,14 @@
-target_include_directories(pxrmp_qed INTERFACE src/physics/quantum_sync)
+find_package(Boost)
+if(Boost_FOUND)
+    include_directories(${Boost_INCLUDE_DIRS})
```
Header-only libs and math libs:
```suggestion
    target_link_libraries(pxrmp_qed INTERFACE Boost::Boost Boost::math)
```

lucafedeli88 committed:
Update src/multi_physics/CMakeLists.txt

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

lucafedeli88 committed:
Update src/multi_physics/QED_tests/CMakeLists.txt

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

lucafedeli88 committed:
Update src/multi_physics/QED_tests/CMakeLists.txt

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

lucafedeli88 committed:
Update src/multi_physics/QED/CMakeLists.txt

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

lucafedeli88 committed:
Update src/multi_physics/QED/CMakeLists.txt

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

lucafedeli88 committed:
Update src/multi_physics/QED_tests/CMakeLists.txt

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

lucafedeli88 committed:
added minimum version

lucafedeli88 committed:
add HAS_BOOST flag

lucafedeli88 committed:
Update src/multi_physics/QED_tests/CMakeLists.txt

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

lucafedeli88 committed:
add missing PRIVATE

lucafedeli88 commented:
@ax3l , thanks really a lot for all the suggestions!

lucafedeli88 commented:
I think that we should use ``find_package(Boost REQUIRED)` instead of `find_package(Boost REQUIRED COMPONENTS math)` : https://gitlab.kitware.com/cmake/cmake/-/issues/18890

ax3l commented on src/multi_physics/QED/CMakeLists.txt:
```diff
@@ -1,19 +1,14 @@
-target_include_directories(pxrmp_qed INTERFACE src/physics/schwinger)
-target_include_directories(pxrmp_qed INTERFACE src/physics/breit_wheeler)
-target_include_directories(pxrmp_qed INTERFACE src/physics/quantum_sync)
+find_package(Boost 1.65.1 COMPONENTS math)
+if(Boost_FOUND)
+    target_link_libraries(pxrmp_qed INTERFACE Boost::Boost Boost::math)
```
```suggestion
find_package(Boost 1.65.1)
if(Boost_FOUND)
    target_link_libraries(pxrmp_qed INTERFACE Boost::boost)
```

ax3l committed:
Multi-Physics: Add CI

ax3l committed:
Boost: Try without math

ax3l committed:
macOS CI: Install Boost

ax3l commented on src/multi_physics/CMakeLists.txt:
```diff
@@ -1,13 +1,33 @@
 
-set(CMAKE_CXX_EXTENSIONS OFF)
+project(
+  pxrmp_qed
```
```suggestion
  pxrmp
```

ax3l committed:
CMake: Move QED target

ax3l committed:
Move QED Tests

ax3l committed:
Remove Wshadow warning

ax3l committed:
Boost: 1.66.0+

The header `boost/math/quadrature/tanh_sinh.hpp` was introduced in
this release.

lucafedeli88 committed:
Merge remote-tracking branch 'upstream/development' into improve_makefile

lucafedeli88 committed:
add ifdefs to remove openMP if needed

lucafedeli88 committed:
option to have boost without table generation support

lucafedeli88 committed:
enable warnings

lucafedeli88 committed:
remove warning in test

lucafedeli88 committed:
fixed warning

lucafedeli88 committed:
moved options to top CMakeLists.txt

ax3l commented on src/multi_physics/CMakeLists.txt:
```diff
@@ -1,13 +1,23 @@
+  PXRMP
+  DESCRIPTION
+    "A header only c++ library (part of the PICSAR library) providing QED modules for Particle-In-Cell codes."
+  HOMEPAGE_URL "https://github.com/ECP-WarpX/picsar"
+  LANGUAGES CXX)
```
Let's avoid this line, since practically we do also depend on the CUDA language when using the interface libraries herein:
```suggestion
  HOMEPAGE_URL "https://github.com/ECP-WarpX/picsar")
```

Activation of CUDA is a bit intricate with AMReX, so let's avoid it here.

ax3l commented on src/multi_physics/QED/CMakeLists.txt:
```diff
@@ -1,19 +1,37 @@
+add_library(${PROJECT_NAME}::QED ALIAS QED)
 
-set(CMAKE_CXX_EXTENSIONS OFF)
+target_compile_features(QED INTERFACE cxx_std_14)
```
Let us try this, because this might be changing the language of the files:
```suggestion
if(CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME)
    target_compile_features(QED INTERFACE cxx_std_14)
endif()
```

lucafedeli88 committed:
Update src/multi_physics/CMakeLists.txt

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

lucafedeli88 committed:
fixed CMakeLists for table_generator

lucafedeli88 committed:
fixed warning

lucafedeli88 committed:
now std:: math functions are used by default

lucafedeli88 committed:
Merge remote-tracking branch 'upstream/development' into improve_makefile

lucafedeli88 committed:
fix issues related to double to float conversions

lucafedeli88 committed:
fixed merge

lucafedeli88 committed:
Merge remote-tracking branch 'upstream/development' into improve_makefile

ax3l commented on src/multi_physics/QED/CMakeLists.txt:
```diff
@@ -1,19 +1,44 @@
+    endif()
+endif()
+
+target_include_directories(QED INTERFACE src/)
```
I would potentially skipp all but the `src/` include and add a prefix `picsar/` generally inside it to avoid collisions with third party libraries.

ax3l commented on src/multi_physics/QED/CMakeLists.txt:
```diff
@@ -1,19 +1,44 @@
+
+target_include_directories(QED INTERFACE src/)
+target_include_directories(QED INTERFACE src/containers)
+target_include_directories(QED INTERFACE src/math)
```
This one definitely collides with includes for CUDA constants from the CUDA Toolkit:
```suggestion
```

ax3l committed:
Do not include `math/` in -I

ax3l commented on src/multi_physics/QED/tests/CMakeLists.txt:
```diff
@@ -0,0 +1,67 @@
+    set_target_properties(${name} PROPERTIES
+        RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin)
+
+    # Enforce C++14
```
```suggestion
    # Require C++14 or newer
```

lucafedeli88 committed:
includes now use full path starting from root

lucafedeli88 committed:
large reorganization

lucafedeli88 committed:
updated automated tests

lucafedeli88 committed:
updated automated tests

lucafedeli88 committed:
removed line in dependencies_mac.sh

lucafedeli88 committed:
modified dependencies_mac.sh

lucafedeli88 committed:
modified dependencies_mac.sh

lucafedeli88 committed:
Merge remote-tracking branch 'upstream/development' into improve_makefile

lucafedeli88 committed:
removed debug line

lucafedeli88 committed:
sycl fix (hopefully)

lucafedeli88 committed:
added fix for dpc++

lucafedeli88 committed:
fixed bug

lucafedeli88 committed:
added comment

lucafedeli88 committed:
fixed bug

lucafedeli88 committed:
sycl fix

lucafedeli88 committed:
changed logic for FORCE_BOOST for special functions

lucafedeli88 committed:
updated README

lucafedeli88 committed:
updated gitignore

NeilZaim commented on .github/workflows/dependencies/dependencies.sh:
```diff
@@ -0,0 +1,18 @@
+#!/usr/bin/env bash
+#
+# Copyright 2020 The PICSAR Community
```
Not sure how important are these copyrights but does it matter if it's `The PICSAR Community` in this file and `The AMReX Community` in the other one (`dependencies_mac.sh`)? @ax3l

lucafedeli88 replied:
Uhm... I don't know which is the policy in this case. @ax3l , what do you think?

ax3l replied:
I usually update them once a year and bump the upper range, usually in January. But practically you can also update when you last touch the file. Both is common. Disclaimer: IANAL ;-)

You can do this in a follow up, I usually do one swipe like this: https://github.com/openPMD/openPMD-api/pull/922

NeilZaim commented on multi_physics/QED/CMakeLists.txt:
```diff
@@ -0,0 +1,85 @@
+project(
+  PXRMP_QED
+  DESCRIPTION
+    "A header only c++ library (part of the PICSAR library) providing QED modules for Particle-In-Cell codes."
```
```suggestion
    "A header only C++ library (part of the PICSAR library) providing QED modules for Particle-In-Cell codes."
```

NeilZaim commented on multi_physics/QED/CMakeLists.txt:
```diff
@@ -0,0 +1,85 @@
+    if(NOT PXRMP_QED_TABLEGEN)
+         message( FATAL_ERROR "Tools require to enable table generation!" )
+    endif()
+  add_subdirectory(QED_table_generator)
```
```suggestion
    add_subdirectory(QED_table_generator)
```

NeilZaim commented on multi_physics/QED/QED_table_generator/CMakeLists.txt:
```diff
@@ -0,0 +1,31 @@
+
+target_link_libraries(${name} PRIVATE PXRMP_QED)
+
+#OpenMP support support
```
```suggestion
# OpenMP support
```

NeilZaim commented on multi_physics/QED/QED_table_generator/CMakeLists.txt:
```diff
@@ -0,0 +1,31 @@
+set_target_properties(${name} PROPERTIES
+    RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/tools)
+
+# Enforce C++14
```
```suggestion
# Require C++14 or newer
```

Following Axel's previous suggestion

NeilZaim commented on multi_physics/QED/QED_tests/CMakeLists.txt:
```diff
@@ -0,0 +1,69 @@
+    set_target_properties(${name} PROPERTIES
+        RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin)
+
+    # Enforce C++14
```
```suggestion
    # Require C++14 or newer
```
Following Axel's previous suggestion

NeilZaim commented on multi_physics/QED/include/picsar_qed/math/cmath_overloads.hpp:
```diff
@@ -99,7 +103,11 @@ namespace math{
     RealType m_floor(const RealType x) noexcept
     {
+#if defined(PXRMP_INTERNAL_DPCPP_FIX) && defined(__SYCL_DEVICE_ONLY__)
+    return cl::sycl::floor(x);
```
```suggestion
        return cl::sycl::floor(x);
```

NeilZaim commented on multi_physics/QED/include/picsar_qed/math/cmath_overloads.hpp:
```diff
@@ -306,7 +314,11 @@ namespace math{
     float m_floor(const float x) noexcept
     {
+#if defined(PXRMP_INTERNAL_DPCPP_FIX) && defined(__SYCL_DEVICE_ONLY__)
+            return cl::sycl::floorf(x);
```
```suggestion
        return cl::sycl::floorf(x);
```

NeilZaim commented on multi_physics/QED/include/picsar_qed/physics/breit_wheeler/breit_wheeler_engine_tables_generator.hpp:
```diff
@@ -181,7 +186,9 @@ namespace breit_wheeler{
         int count = 0;
+    #ifdef PXRMP_HAS_OPENMP
         #pragma omp parallel for schedule(dynamic, 1)
+    #endif
```
```suggestion
#ifdef PXRMP_HAS_OPENMP
        #pragma omp parallel for schedule(dynamic, 1)
#endif
```

NeilZaim commented on multi_physics/QED/include/picsar_qed/physics/quantum_sync/quantum_sync_engine_tables_generator.hpp:
```diff
@@ -179,7 +184,9 @@ namespace quantum_sync{
         int count = 0;
+    #ifdef PXRMP_HAS_OPENMP
         #pragma omp parallel for schedule(dynamic, 1)
+    #endif
```
```suggestion
#ifdef PXRMP_HAS_OPENMP
        #pragma omp parallel for schedule(dynamic, 1)
#endif
```

NeilZaim commented on shippable.yml:
```diff
@@ -16,7 +16,7 @@ build:
 
     #QED LIBRARY
-    - cd /home/warp_user/src/bitbucket.org/berkeleylab/picsar/src/multi_physics
+    - cd /home/warp_user/src/bitbucket.org/berkeleylab/picsar/multi_physics/QED
```
We're currently not using shippable right? Should we remove this file altogether from `development`?

lucafedeli88 replied:
Not that I am aware of ! However, we might see this as an opportunity to start using LGTM with PICSAR-QED (since I think that LGTM needs this file to work). LGTM should be free for open source software, right? What do you think @NeilZaim  and @ax3l  ?

ax3l replied:
Just closing the loop from slack: I think LGTM is now integrated here.

NeilZaim reviewed (commented):
I'm really not familiar with CMake so I only have very minor (e.g. formatting) comments. I'd let @ax3l decide if this PR is ready for merging.

lucafedeli88 committed:
Update multi_physics/QED/QED_table_generator/CMakeLists.txt

Co-authored-by: NeilZaim <49716072+NeilZaim@users.noreply.github.com>

lucafedeli88 committed:
Update multi_physics/QED/CMakeLists.txt

Co-authored-by: NeilZaim <49716072+NeilZaim@users.noreply.github.com>

lucafedeli88 committed:
Update multi_physics/QED/QED_table_generator/CMakeLists.txt

Co-authored-by: NeilZaim <49716072+NeilZaim@users.noreply.github.com>

lucafedeli88 committed:
Update multi_physics/QED/include/picsar_qed/physics/quantum_sync/quantum_sync_engine_tables_generator.hpp

Co-authored-by: NeilZaim <49716072+NeilZaim@users.noreply.github.com>

lucafedeli88 committed:
Update multi_physics/QED/include/picsar_qed/math/cmath_overloads.hpp

Co-authored-by: NeilZaim <49716072+NeilZaim@users.noreply.github.com>

lucafedeli88 committed:
Update multi_physics/QED/include/picsar_qed/math/cmath_overloads.hpp

Co-authored-by: NeilZaim <49716072+NeilZaim@users.noreply.github.com>

lucafedeli88 committed:
Update multi_physics/QED/include/picsar_qed/physics/breit_wheeler/breit_wheeler_engine_tables_generator.hpp

Co-authored-by: NeilZaim <49716072+NeilZaim@users.noreply.github.com>

lucafedeli88 committed:
Update multi_physics/QED/QED_tests/CMakeLists.txt

Co-authored-by: NeilZaim <49716072+NeilZaim@users.noreply.github.com>

lucafedeli88 committed:
Update multi_physics/QED/CMakeLists.txt

Co-authored-by: NeilZaim <49716072+NeilZaim@users.noreply.github.com>

lucafedeli88 committed:
not using INTERNAL preprocessor variables anymore

lucafedeli88 committed:
try to add .lgtm.yml file

lucafedeli88 committed:
trying to add codeql-analysis.yml

lucafedeli88 committed:
improve codeql-analysis.yml

lucafedeli88 committed:
bugfixing in codeql-analysis.yml

lucafedeli88 committed:
analyze only C++

lucafedeli88 committed:
do not run tests while doing code analysis

lucafedeli88 committed:
updated lgtm file

lucafedeli88 commented:
I think that once we take a decision on the copyright in `dependencies.sh` and on removing  the `shippable`  file this PR can be merged!

NeilZaim commented on multi_physics/QED/include/picsar_qed/qed_commons.h:
```diff
@@ -79,57 +86,59 @@
- * implementation of the upper_bound algorithm.
- * Otherwise, a choice is made based on if GPU
- * support is enabled or not.
+ * If GPU support is enables, the library uses an internal, GPU-friendly,
```
```suggestion
 * If GPU support is enabled, the library uses an internal, GPU-friendly,
```

lucafedeli88 committed:
Update multi_physics/QED/include/picsar_qed/qed_commons.h

Co-authored-by: NeilZaim <49716072+NeilZaim@users.noreply.github.com>

ax3l commented on multi_physics/QED/QED_table_generator/table_generator.cpp:
```diff
@@ -9,12 +9,15 @@
+    #include <omp.h>
+#endif
+
+#include <picsar_qed/physics/quantum_sync/quantum_sync_engine_tables.hpp>
+#include <picsar_qed/physics/quantum_sync/quantum_sync_engine_tables_generator.hpp>
+#include <picsar_qed/physics/quantum_sync/quantum_sync_engine_tabulated_functions.hpp>
+#include <picsar_qed/physics/breit_wheeler/breit_wheeler_engine_tables_generator.hpp>
```
Uh, the internal includes here should actually use `"..."` to avoid prefering an externally installed, potentially older version of picsar.

```suggestion
#include "picsar_qed/physics/quantum_sync/quantum_sync_engine_tables.hpp"
#include "picsar_qed/physics/quantum_sync/quantum_sync_engine_tables_generator.hpp"
#include "picsar_qed/physics/quantum_sync/quantum_sync_engine_tabulated_functions.hpp"
#include "picsar_qed/physics/breit_wheeler/breit_wheeler_engine_tables_generator.hpp"
```

lucafedeli88 replied:
Ah, sure! thanks!

ax3l commented on multi_physics/QED/QED_tests/test_picsar_algo.cpp:
```diff
@@ -11,8 +11,8 @@
 #include <algorithm>
```
usually include stdlib includes last, so you notice quickly when an include is missing in a header file

lucafedeli88 replied:
I have just rearranged the stdlib includes in all the tests!

ax3l commented on multi_physics/QED/include/picsar_qed/qed_commons.h:
```diff
@@ -0,0 +1,144 @@
+ * provided by Boost are used.
+ */
+ #ifdef PXRMP_USE_CXX17_FOR_SPECIAL_FUNCTIONS
+     #if !(__cplusplus > 201402L)
```
You mean
```suggestion
     #if __cplusplus < 201703L
```
https://gist.github.com/ax3l/53db9fa8a4f4c21ecc5c4100c0d93c94#meaning-of-__cplusplus

lucafedeli88 replied:
yeah, you solution is better

ax3l commented on multi_physics/QED/include/picsar_qed/qed_commons.h:
```diff
@@ -0,0 +1,144 @@
+ * picsar makes use of "if constexpr". Otherwise, the
+ * expression falls back to a regular "if".
+ */
+  #if __cplusplus > 201402L
```
Also just slightly more precise
```suggestion
  #if __cplusplus >= 201703L
```

ax3l commented on multi_physics/QED/include/picsar_qed/qed_commons.h:
```diff
@@ -0,0 +1,144 @@
+/**
+* PXRMP_DPCPP_FIX enables a workaround to allow compilation with
+* DPC++, which apparently has issues with floorf or std::floor(x) when
+* x is a float
```
Let's open an issue about this at https://github.com/intel/llvm and link it here for tracking

lucafedeli88 replied:
@ax3l , I think that this is a know issue: https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/C-CXX-StandardLibrary/C-CXX-StandardLibrary.rst . In the documentation they explicitly state that not all `math.h` functions are supported right now, including `floor` and `floorf` . Should we open an issue anyway?

lucafedeli88 committed:
Update multi_physics/QED/include/picsar_qed/qed_commons.h

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

lucafedeli88 committed:
Update multi_physics/QED/QED_table_generator/table_generator.cpp

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

lucafedeli88 committed:
rearranged c++ stdlib includes

lucafedeli88 committed:
Update multi_physics/QED/include/picsar_qed/qed_commons.h

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>
